operator/mysqlcontroller/observe.go

```diff
@@ -0,0 +1,180 @@
+func (p *Pipeline) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	log := controllerruntime.LoggerFrom(ctx)
+	log.V(1).Info("observing resource")
+
+	mySQLInstance := mg.(*exoscalev1.MySQL)
+
+	resp, err := p.exo.GetDbaasServiceMysqlWithResponse(ctx, oapi.DbaasServiceName(mySQLInstance.GetInstanceName()))
+	if err != nil {
+		if errors.Is(err, exoscaleapi.ErrNotFound) {
+			return managed.ExternalObservation{ResourceExists: false}, nil
+		}
+		return managed.ExternalObservation{}, fmt.Errorf("cannot observe mySQLInstance: %w", err)
+	}
+
+	mysql := *resp.JSON200
+	log.V(2).Info("response", "raw", string(resp.Body))
+	log.V(1).Info("retrieved mySQLInstance", "state", mysql.State)
+
+	mySQLInstance.Status.AtProvider, err = mapObservation(mysql)
+	if err != nil {
+		log.Error(err, "cannot map mySQLInstance observation, ignoring")
+	}
+	var state oapi.EnumServiceState
+	if mysql.State != nil {
+		state = *mysql.State
+	}
+	switch state {
+	case oapi.EnumServiceStateRunning:
+		mySQLInstance.SetConditions(exoscalev1.Running())
+	case oapi.EnumServiceStateRebuilding:
+		mySQLInstance.SetConditions(exoscalev1.Rebuilding())
+	case oapi.EnumServiceStatePoweroff:
+		mySQLInstance.SetConditions(exoscalev1.PoweredOff())
+	case oapi.EnumServiceStateRebalancing:
+		mySQLInstance.SetConditions(exoscalev1.Rebalancing())
+	default:
+		log.V(2).Info("ignoring unknown mySQLInstance state", "state", state)
+	}
+
+	caCert, err := p.exo.GetDatabaseCACertificate(ctx, mySQLInstance.Spec.ForProvider.Zone)
+	if err != nil {
+		return managed.ExternalObservation{}, fmt.Errorf("cannot retrieve CA certificate: %w", err)
+	}
+
+	connDetails, err := connectionDetails(mysql, caCert)
+	if err != nil {
+		return managed.ExternalObservation{}, fmt.Errorf("cannot parse connection details: %w", err)
+	}
+
+	params, err := mapParameters(mysql, mySQLInstance.Spec.ForProvider.Zone)
+	if err != nil {
+		return managed.ExternalObservation{}, fmt.Errorf("cannot parse parameters: %w", err)
+	}
+
+	return managed.ExternalObservation{
+		ResourceExists:    true,
+		ResourceUpToDate:  isUpToDate(&mySQLInstance.Spec.ForProvider, params, log),
+		ConnectionDetails: connDetails,
+	}, nil
+}
+
+func connectionDetails(in oapi.DbaasServiceMysql, ca string) (managed.ConnectionDetails, error) {
+	uri := pointer.StringDeref(in.Uri, "")
+	// uri may be absent
+	if uri == "" {
+		if in.ConnectionInfo == nil || in.ConnectionInfo.Uri == nil || len(*in.ConnectionInfo.Uri) == 0 {
+			return map[string][]byte{}, nil
+		}
+		uri = (*in.ConnectionInfo.Uri)[0]
+	}
+	parsed, err := url.Parse(uri)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse connection URI: %w", err)
+	}
+	password, _ := parsed.User.Password()
+	return map[string][]byte{
+		"MYSQL_USER":     []byte(parsed.User.Username()),
+		"MYSQL_PASSWORD": []byte(password),
+		"MYSQL_URL":      []byte(uri),
+		"MYSQL_DB":       []byte(strings.TrimPrefix(parsed.Path, "/")),
```

Comment:
Maybe write it out fully if there's a convention for mysql client?
```suggestion
		"MYSQL_DATABASE":       []byte(strings.TrimPrefix(parsed.Path, "/")),
```